File: ComputationalGeometry/ApplicationOfTriangulation/myLibraries/util/geometry/DCEL/MonotoneVertex.java
package myLibraries.util.geometry.DCEL;

/*
 * MonotoneVertex.java
 *
 * Version:
 *     $1.0$
 *
 * Revisions:
 *     $1.0$
 */

import myLibraries.util.geometry.elements.point.Vector;
import myLibraries.util.geometry.tools.Triangles;
import myLibraries.lang.MyMath;

/**
 * Data structure of Monotone Vertex
 * for partitioning monotone subpolygon
 *
 * @author       Xiaoyu Tongyang, or call me sora for short
 */

public class MonotoneVertex extends Vertex {
    public static final boolean LEFT_CHAIN_VERTEX = true;
    public static final boolean RIGHT_CHAIN_VERTEX = false;
    public boolean isLeftChainVertex;

    public VertexType vertexType;

    /**
     * enumerative Vertex Type for partitioning monotone subpolygons
     * */

    public enum VertexType {
        START, SPLIT, // 0, 1
        END, MERGE, // 2, 3
        REGULAR_LEFT, REGULAR_RIGHT, // 4
    }

    /**
     * constructs to create an instance of MonotoneVertex
     * */

    public MonotoneVertex( double x, double y ) {
        super( x, y );
    }

    /**
     * this vertex is a split or merge one
     * when the angle it forms is greater than pi.
     * */

    public boolean isSplitOrMergeVertex() {
        return MyMath.isSmallerThanZero( 
                Triangles.areaTwo( incidentEdge.prev.origin, incidentEdge.origin,
                        incidentEdge.next.origin ) );
    }

    /**
     * are both vertices on the same monotone chain?
     * */

    public boolean isOnTheDifferentChain( MonotoneVertex vertex ) {
        return isLeftChainVertex != vertex.isLeftChainVertex;
    }

    public static
    void main( String[] args ) {
        Vector vector1 = new Vector( -5, -5 );
        Vector vector2 = new Vector( 0, 3 );
        Vector vector3 = new Vector( 2, -1 );
        System.out.println( Triangles.areaTwo( vector1, vector2, vector3 ) );
    }
}

File: web/myLibraries/util/geometry/DCEL/MonotoneVertex.js
"use strict"

/*
 * MonotoneVertex.java
 *
 * Version:
 *     $1.0$
 *
 * Revisions:
 *     $1.0$
 */

import Vertex from "./Vertex.js";
import Triangles from "../tools/Triangles.js";
import MyMath from "../../../lang/MyMath.js";

/**
 * Data structure of Monotone Vertex
 * for partitioning monotone subpolygon
 *
 * The source code in java is from my own github:
 * @see <a href=https://github.com/fengkeyleaf/Algorithm/blob/main/ComputationalGeometry/ApplicationOfTriangulation/myLibraries/util/geometry/DCEL/MonotoneVertex.java>MonotoneVertex</a>
 *
 * @author       Xiaoyu Tongyang, or call me sora for short
 */

export default class MonotoneVertex extends Vertex {
    static LEFT_CHAIN_VERTEX = true;
    static RIGHT_CHAIN_VERTEX = false;

    static LEFT_CHAIN_VERTEX_COLOR = [ 255 / 255, 0, 102 / 255, 1 ];
    static RIGHT_CHAIN_VERTEX_COLOR = [ 0, 204 / 255, 255 / 255, 1 ];

    /**
     * enumerative Vertex Type for partitioning monotone subpolygons
     * */

    static VertexType = {
        START: 0, SPLIT: 1, // 0, 1
        END: 2, MERGE: 3, // 2, 3
        REGULAR_LEFT: 4, REGULAR_RIGHT: 5, // 4
    };

    /**
     * constructs to create an instance of MonotoneVertex
     *
     * @param {Number} x
     * @param {Number} y
     * */

    constructor( x, y ) {
        super( x, y );
        this.isLeftChainVertex = false;
        this.vertexType = -1;
    }

    /**
     * this vertex is a split or merge one
     * when the angle it forms is greater than pi.
     * */

    isSplitOrMergeVertex() {
        return MyMath.isNegative(
            Triangles.areaTwo( this.incidentEdge.prev.origin, this.incidentEdge.origin,
                this.incidentEdge.next.origin ) );
    }

    /**
     * are both vertices on the same monotone chain?
     *
     * @param {MonotoneVertex} vertex
     * */

    isOnTheDifferentChain( vertex ) {
        return this.isLeftChainVertex !== vertex.isLeftChainVertex;
    }

}
